File: backend/src/InsuraTech.Application/Policies/Commands/CreatePolicy/Strategies/CreateHealthPolicyStrategy.cs
using InsuraTech.Domain.Policies;
using InsuraTech.Domain.Policies.ValueObjects;

namespace InsuraTech.Application.Policies.Commands.CreatePolicy.Strategies;

public sealed class CreateHealthPolicyStrategy : ICreatePolicyStrategy
{
    public PolicyType Type => PolicyType.Health;

    public bool CanHandle(CreatePolicyCommand request) =>
        !string.IsNullOrWhiteSpace(request.HealthPlanId);

    public Task<Policy> CreateAsync(CreatePolicyCommand request, PolicyNumber policyNumber, InsuredPerson insured, CancellationToken cancellationToken)
    {
        var coverage = CoveragePeriod.Create(request.CoverageStartDate, request.CoverageEndDate);
        var policy   = Policy.CreateHealthPolicy(
            policyNumber, insured, coverage,
            request.MonthlyPremium, request.HealthPlanId!,
            DateOnly.FromDateTime(DateTime.UtcNow));

        return Task.FromResult(policy);
    }
}

File: backend_v2/src/main/java/com/insuratech/application/policies/strategies/CreateHealthPolicyStrategy.java
// Spec: SPEC-007 — health-plan
// Origen: CreateHealthPolicyStrategy.cs → CreateHealthPolicyStrategy.java
package com.insuratech.application.policies.strategies;

import com.insuratech.application.policies.commands.CreatePolicyCommand;
import com.insuratech.domain.policies.Policy;
import com.insuratech.domain.policies.PolicyType;
import com.insuratech.domain.policies.health.HealthPlanCatalog;
import com.insuratech.domain.policies.health.HealthPlanSelection;
import com.insuratech.domain.policies.vo.CoveragePeriod;
import com.insuratech.domain.policies.vo.InsuredPerson;
import com.insuratech.domain.policies.vo.PolicyNumber;
import org.springframework.stereotype.Component;

@Component
public class CreateHealthPolicyStrategy implements PolicyCreationStrategy {

    @Override
    public PolicyType type() { return PolicyType.HEALTH; }

    @Override
    public boolean canHandle(CreatePolicyCommand cmd) {
        return cmd.healthPlanId() != null && !cmd.healthPlanId().isBlank();
    }

    @Override
    public Policy create(CreatePolicyCommand cmd, PolicyNumber policyNumber, InsuredPerson insured) {
        var plan = HealthPlanCatalog.findById(cmd.healthPlanId())
            .orElseThrow(() -> new IllegalArgumentException("Health plan not found: " + cmd.healthPlanId()));
        var selection = new HealthPlanSelection(plan, insured.getAge());
        var period = CoveragePeriod.of(cmd.coverageStartDate(), cmd.coverageEndDate());
        return Policy.createHealthPolicy(policyNumber, insured, period, selection, cmd.agentId(), cmd.cityId());
    }
}
